util/log: overwrite log tags with tags from an AmbientContext

andreimatei opened the pull request:
Today, annotating a context doesn't overwrite existing tags. We have
tests in which a replica uses its context to call into another replica,
and it's very confusing that those log messages appear to be in the
context of the caller replica.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/12515)
<!-- Reviewable:end -->

andreimatei committed:
util/log: overwrite log tags with tags from an AmbientContext

Today, annotating a context doesn't overwrite existing tags. We have
tests in which a replica uses its context to call into another replica,
and it's very confusing that those log messages appear to be in the
context of the caller replica.

tamird commented:
Reviewed 2 of 4 files at r1.
Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/util/log/log_context_test.go, line 146 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515#-KZTImSdhUgZhs_lF7im:-KZTImSdhUgZhs_lF7in:bsbt9rp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c84f1181ca3ac693758340df36d94b1abe9f02/pkg/util/log/log_context_test.go#L146)):*
> ```Go
> 	}
> 
> 	for i, tc := range testCases {
> ```

kill the loop index if you're going to use subtests plz

leaving the meat of this review to radu.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 2 of 4 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/util/log/log_context.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515#-KZTtuztHLr_0hhtT3Iv:-KZTtuzuAYc1zu75hmrl:bwvjac2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c84f1181ca3ac693758340df36d94b1abe9f02/pkg/util/log/log_context.go#L30)):*
> ```Go
> // messages under that context.
> //
> // The logTag entries form a linked chain (newer (a.k.a bottom) to older),
> ```

`- newer (aka bottom) to older -`

---

*[pkg/util/log/log_context.go, line 151 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515#-KZTw0CwBHfphR7_F4bI:-KZTw0CwBHfphR7_F4bJ:b-276i5a) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c84f1181ca3ac693758340df36d94b1abe9f02/pkg/util/log/log_context.go#L151)):*
> ```Go
> 
> // copyChain takes (the head of) a chain, copies all the nodes in the chain, and
> // returns the head and tail of the copy.
> ```

might be better to only use `top` and `bottom`. Head and tail are technically correct but they're counterintuitive because top = tail and bottom = head

---

*[pkg/util/log/log_context.go, line 176 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515#-KZTum9Q9LVABsMoTtvd:-KZTum9R4NKz4DX1S97p:b-kycf4k) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c84f1181ca3ac693758340df36d94b1abe9f02/pkg/util/log/log_context.go#L176)):*
> ```Go
> // are copied.
> func subtractChain(chain *logTag, vals *logTag) *logTag {
> 	// nodes will contain the elements of the returned chain.
> ```

We should have a fast path that doesn't allocate anything if there are no duplicate tags. In fact, we should always avoid copying the topmost tags that aren't duplicates.

We could pass to `copyChain` the top's parent so it's able to copy subchains and use that for every contiguous piece of non-duplicate tags. Something like:

```
copyTop, copyBottom = nil, nil
for t != nil {
  // Skip any duplicate tags.
  for t != nil && chainContains(t) { t = t.parent }
  if t == nil { break }
  subchainBottom, subchainTop = t, t
  for t.parent != nil && !chainContains(t.parent) { subchainTop, t = t, t.parent }
  // We have a subchain of non-duplicate tags. If it ends at the top, we can use it directly;
  // otherwise we need to make a copy because we will reattach it to another parent.
  if t != nil {
     subchainBottom, subchainTop = copyChain(subchainBottom, t)
  }
  if copyTop != nil {
     copyTop.parent = subchainBottom
  } else {
     copyBottom = subchainBottom
  }
 copyTop = subchainTop
}
return copyBottom
```

---

*[pkg/util/log/log_context.go, line 206 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515#-KZTuL2LozL6J5j6v3Lx:-KZTuL2LozL6J5j6v3Ly:b-d1yho9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c84f1181ca3ac693758340df36d94b1abe9f02/pkg/util/log/log_context.go#L206)):*
> ```Go
> }
> 
> func chainContains(vals *logTag, key string) bool {
> ```

[nit] could you move it up before `subtractChain` which uses it

---

*[pkg/util/log/log_context_test.go, line 147 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515#-KZTsv8HZmFVN1wRYyoK:-KZTsv8INVBPwGOPHztj:bwir4h6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c84f1181ca3ac693758340df36d94b1abe9f02/pkg/util/log/log_context_test.go#L147)):*
> ```Go
> 
> 	for i, tc := range testCases {
> 		t.Run("", func(t *testing.T) {
> ```

you can use `t.expected` as the subtest name. you can remove it from the error message below

---

*[pkg/util/log/log_context_test.go, line 159 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515#-KZTt8uOrGJ2I8Kfeu-n:-KZTt8uPFQRp_rAR7XpC:btj83o2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c84f1181ca3ac693758340df36d94b1abe9f02/pkg/util/log/log_context_test.go#L159)):*
> ```Go
> }
> 
> func (c *chain) Append(t logTag) *chain {
> ```

unexport

---

*[pkg/util/log/log_context_test.go, line 163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515#-KZTtEv0vcIy5ieKVIdE:-KZTtEv0vcIy5ieKVIdF:bs5e7ia) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c84f1181ca3ac693758340df36d94b1abe9f02/pkg/util/log/log_context_test.go#L163)):*
> ```Go
> 		panic("can't append a chain")
> 	}
> 	if c.head == nil {
> ```

pull out `c.tail = &t`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 2 of 4 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/util/log/log_context.go, line 176 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515#-KZTum9Q9LVABsMoTtvd:-KZU-dplfs1DVB4QGg1z:b5jk12y) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c84f1181ca3ac693758340df36d94b1abe9f02/pkg/util/log/log_context.go#L176)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We should have a fast path that doesn't allocate anything if there are no duplicate tags. In fact, we should always avoid copying the topmost tags that aren't duplicates.

We could pass to `copyChain` the top's parent so it's able to copy subchains and use that for every contiguous piece of non-duplicate tags. Something like:

```
copyTop, copyBottom = nil, nil
for t != nil {
  // Skip any duplicate tags.
  for t != nil && chainContains(t) { t = t.parent }
  if t == nil { break }
  subchainBottom, subchainTop = t, t
  for t.parent != nil && !chainContains(t.parent) { subchainTop, t = t, t.parent }
  // We have a subchain of non-duplicate tags. If it ends at the top, we can use it directly;
  // otherwise we need to make a copy because we will reattach it to another parent.
  if t != nil {
     subchainBottom, subchainTop = copyChain(subchainBottom, t)
  }
  if copyTop != nil {
     copyTop.parent = subchainBottom
  } else {
     copyBottom = subchainBottom
  }
 copyTop = subchainTop
}
return copyBottom
```
</blockquote></details>

Errata: `for t = t.parent; t != nil && !chainContains(t); { subchainTop, t = t, t.parent }`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 2 of 4 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/util/log/log_context.go, line 30 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515#-KZTtuztHLr_0hhtT3Iv:-KZXrT3L7pCBbRj8zean:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c84f1181ca3ac693758340df36d94b1abe9f02/pkg/util/log/log_context.go#L30)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

`- newer (aka bottom) to older -`
</blockquote></details>

Done.

---

*[pkg/util/log/log_context.go, line 151 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515#-KZTw0CwBHfphR7_F4bI:-KZXqEwDkg8nVvucAzuu:b23ql9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c84f1181ca3ac693758340df36d94b1abe9f02/pkg/util/log/log_context.go#L151)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

might be better to only use `top` and `bottom`. Head and tail are technically correct but they're counterintuitive because top = tail and bottom = head
</blockquote></details>

ok, switched to top/bottom everywhere. But I found that confusing - da hell is the "bottom"? I get that it kinda works with contexts, which have "parents"... Anyway, I've added head/tail to the a.k.a's at the top.

---

*[pkg/util/log/log_context.go, line 176 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515#-KZTum9Q9LVABsMoTtvd:-KZXqkDruOsM6fyCbYsI:b414pqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c84f1181ca3ac693758340df36d94b1abe9f02/pkg/util/log/log_context.go#L176)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Errata: `for t = t.parent; t != nil && !chainContains(t); { subchainTop, t = t, t.parent }`
</blockquote></details>

Done. Nice code.

---

*[pkg/util/log/log_context.go, line 206 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515#-KZTuL2LozL6J5j6v3Lx:-KZXqqay5CRH7LGdzokN:bx8oo35) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c84f1181ca3ac693758340df36d94b1abe9f02/pkg/util/log/log_context.go#L206)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] could you move it up before `subtractChain` which uses it
</blockquote></details>

well but this is the better order, isn't it? It's not topological sort, it's higher-level functions first.

---

*[pkg/util/log/log_context_test.go, line 146 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515#-KZTImSdhUgZhs_lF7im:-KZXraxy_q_Olp6nhrwO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c84f1181ca3ac693758340df36d94b1abe9f02/pkg/util/log/log_context_test.go#L146)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

kill the loop index if you're going to use subtests plz

leaving the meat of this review to radu.
</blockquote></details>

Done.

---

*[pkg/util/log/log_context_test.go, line 147 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515#-KZTsv8HZmFVN1wRYyoK:-KZXrhLVd2lq97l1KtrN:b-k9tqux) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c84f1181ca3ac693758340df36d94b1abe9f02/pkg/util/log/log_context_test.go#L147)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

you can use `t.expected` as the subtest name. you can remove it from the error message below
</blockquote></details>

added it as the subtest name, but left it in the error too

---

*[pkg/util/log/log_context_test.go, line 159 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515#-KZTt8uOrGJ2I8Kfeu-n:-KZXs90qrDtVOogbDf63:bwwe274) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c84f1181ca3ac693758340df36d94b1abe9f02/pkg/util/log/log_context_test.go#L159)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

unexport
</blockquote></details>

can't because `append` is reserved. Renamed.

---

*[pkg/util/log/log_context_test.go, line 163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515#-KZTtEv0vcIy5ieKVIdE:-KZXsPo5uKltODAi-ndY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0c84f1181ca3ac693758340df36d94b1abe9f02/pkg/util/log/log_context_test.go#L163)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

pull out `c.tail = &t`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Review status: 2 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12515#-:-KZXwJJmynYhp4pks3qH:b-uwkowd)*
<!-- Sent from Reviewable.io -->
